File: sdk/python/tap/protocol/signing.py
"""Ed25519 signing and verification of `CommitMessage`.

This module is intentionally thin: the only crypto primitive TAP needs at
the protocol layer is Ed25519 over a 60-byte message. Anything beyond that
(key management, session-key lifecycle) lives in the consumer package."""

from __future__ import annotations

from nacl.exceptions import BadSignatureError
from nacl.signing import SigningKey, VerifyKey
from solders.pubkey import Pubkey

from tap.exceptions import CommitmentError
from tap.protocol.codec import encode_commitment_bytes
from tap.protocol.commit import CommitMessage, SignedCommitment


def sign_commitment(message: CommitMessage, signing_key: SigningKey) -> SignedCommitment:
    """Produce a `SignedCommitment` that the on-chain program will accept,
    given the session-key `SigningKey` registered at `open_channel`."""
    sig = signing_key.sign(encode_commitment_bytes(message)).signature
    return SignedCommitment(message=message, signature=sig)


def verify_commitment(signed: SignedCommitment, session_key: Pubkey) -> None:
    """Raise `CommitmentError` if `signed` is not a valid signature by
    `session_key` over `signed.message`."""
    try:
        VerifyKey(bytes(session_key)).verify(
            encode_commitment_bytes(signed.message),
            signed.signature,
        )
    except BadSignatureError as exc:
        raise CommitmentError("commitment signature failed verification") from exc

File: sdk/typescript/src/protocol/signing.ts
/**
 * Ed25519 signing and verification of `CommitMessage`. Thin by design — the
 * only crypto primitive TAP needs at the protocol layer is Ed25519 over the
 * 60-byte canonical message form. Key management lives in the consumer
 * package; this file is pure crypto.
 */

import * as ed from "@noble/ed25519";
import { sha512 } from "@noble/hashes/sha512";

import { CommitmentError } from "../exceptions.js";
import type { CommitMessage, SignedCommitment } from "./commit.js";
import { encodeCommitmentBytes } from "./codec.js";

// `@noble/ed25519` v2 requires a sha-512 implementation to be wired in once
// per process. We use `@noble/hashes/sha512` so the SDK runs unchanged on
// Node, Bun, Deno, and browsers without depending on `node:crypto`.
ed.etc.sha512Sync = (...msgs: Uint8Array[]) => sha512(ed.etc.concatBytes(...msgs));

/** Produce a `SignedCommitment` that the on-chain program will accept, given
 * the 32-byte session-key seed registered at `open_channel`. */
export function signCommitment(
  message: CommitMessage,
  sessionKeySeed: Uint8Array,
): SignedCommitment {
  if (sessionKeySeed.length !== 32) {
    throw new Error("session key seed must be 32 bytes");
  }
  const signature = ed.sign(encodeCommitmentBytes(message), sessionKeySeed);
  return { message, signature };
}

/** Throws `CommitmentError` if `signed` is not a valid signature by
 * `sessionPublicKey` over `signed.message`. */
export function verifyCommitment(
  signed: SignedCommitment,
  sessionPublicKey: Uint8Array,
): void {
  if (sessionPublicKey.length !== 32) {
    throw new CommitmentError("session public key must be 32 bytes");
  }
  const ok = ed.verify(signed.signature, encodeCommitmentBytes(signed.message), sessionPublicKey);
  if (!ok) {
    throw new CommitmentError("commitment signature failed verification");
  }
}

/** Derive the 32-byte Ed25519 public key from a 32-byte secret seed. Exposed
 * separately so consumers can derive the on-chain pubkey from a stored seed
 * without round-tripping through a signer object. */
export function publicKeyFromSeed(sessionKeySeed: Uint8Array): Uint8Array {
  if (sessionKeySeed.length !== 32) {
    throw new Error("session key seed must be 32 bytes");
  }
  return ed.getPublicKey(sessionKeySeed);
}
